### PR TITLE
Parser optimizations

### PR DESCRIPTION
**Pull request by sdkrystian**

### CODE REVIEW

**Comment by sdkrystian** [comment]
This PR was rebased on top of #136, so really only the 5 most recent commits should be merged

File: include/boost/json/detail/basic_parser.hpp [hunk 1]
```diff
@@ -108,59 +100,110 @@ class basic_parser
...
+
+    BOOST_NOINLINE
+    inline
+    std::nullptr_t
+    partial_if_more(
+        const char* p, 
+        state st);
+
+    BOOST_NOINLINE
+    inline
```
Comment (vinniefalco) [inline include/boost/json/detail/basic_parser.hpp:135]: why `BOOST_NOINLINE` followed by `inline` ?

File: include/boost/json/detail/basic_parser.hpp [hunk 2]
```diff
@@ -199,7 +198,8 @@ class basic_parser
     template<bool StackEmpty, class Handler>
     const char* parse_false(Handler& h, const char* p);
     
-    template<bool StackEmpty, bool AllowInvalid, class Handler>
+    template<bool StackEmpty, bool IsKey,
+        bool AllowInvalid, class Handler>
```
Comment (vinniefalco) [inline include/boost/json/detail/basic_parser.hpp:202]: I think we should rename `AllowInvalid` to `AllowBadUTF8`. This change should be in its own commit.

**Comment by vinniefalco** [comment]
I think the "Better key parsing" commit should be removed entirely. It is an optimization, and I do not want to merge any more optimizations, there have been enough already. Any new optimizations should be submitted only when the following conditions have all been met:
1. Boost.JSON is accepted in review,
2. There have been no logic or feature changes merged for at least 60 days, and
3. The change is accompanied by comprehensive benchmarks in a stable environment which demonstrate the benefit.
4. A measurement is provided of the change in the size of the emitted binary with optimizations turned on.

File: include/boost/json/detail/basic_parser.hpp (hunk 1 above)
Comment (sdkrystian) [reply include/boost/json/detail/basic_parser.hpp:135]: We want the ODR benefits of the `inline` keyword (multiple definitions), but we don't actually want the function to be inlined.

**Comment by vinniefalco** [comment]
Too many changes and too many unnecessary changes, sorry
